### PR TITLE
batch_norm momentum support variable

### PR DESCRIPTION
**Pull request by heavengate**
**batch_norm momentum support Variable as input** 
cn doc: https://github.com/PaddlePaddle/FluidDoc/pull/1619
batch_norm1
batch_norm2
batch_norm3

### CODE REVIEW

File: paddle/fluid/operators/batch_norm_op.cc [hunk 1]
```diff
@@ -231,6 +240,12 @@ class BatchNormKernel<platform::CPUDeviceContext, T>
     const DataLayout data_layout =
         framework::StringToDataLayout(data_layout_str);
 
+    float momentum = ctx.Attr<float>("momentum");
+    if (ctx.HasInput("MomentumTensor")) {
+      const auto *mom_tensor = ctx.Input<Tensor>("MomentumTensor");
+      momentum = mom_tensor->data<float>()[0];
+    }
```
Comment (qingqing01) [inline paddle/fluid/operators/batch_norm_op.cc]: put the if branch into the tranin-phase

File: paddle/fluid/operators/batch_norm_op.cu [hunk 2]
```diff
@@ -43,13 +43,21 @@ class BatchNormKernel<platform::CUDADeviceContext, T>
...
     double epsilon = static_cast<double>(ctx.Attr<float>("epsilon"));
-    const float momentum = ctx.Attr<float>("momentum");
     const bool is_test = ctx.Attr<bool>("is_test");
     const bool use_global_stats = ctx.Attr<bool>("use_global_stats");
     const std::string data_layout_str = ctx.Attr<std::string>("data_layout");
     const DataLayout data_layout =
         framework::StringToDataLayout(data_layout_str);
 
+    float momentum = ctx.Attr<float>("momentum");
+    if (ctx.HasInput("MomentumTensor")) {
```
Comment (qingqing01) [inline paddle/fluid/operators/batch_norm_op.cu]: same as above

File: python/paddle/fluid/layers/nn.py [hunk 3]
```diff
@@ -2432,16 +2432,17 @@ def batch_norm(input,
...
         act(string, Default None): Activation type, linear|relu|prelu|...
         is_test (bool, Default False): A flag indicating whether it is in
             test phrase or not.
-        momentum(float, Default 0.9): The value used for the moving_mean and
+        momentum(float|Variable, Default 0.9): The value used for the moving_mean and
             moving_var computation. The updated formula is:
             :math:`moving\_mean = moving\_mean * momentum + new\_mean * (1. - momentum)`
             :math:`moving\_var = moving\_var * momentum + new\_var * (1. - momentum)`
-            Default is 0.9.
+            This should be a float number or a Variable with shape as [1] and data
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: with shape as [1] ->  with shape [1]

File: python/paddle/fluid/layers/nn.py [hunk 4]
```diff
@@ -2432,16 +2432,17 @@ def batch_norm(input,
...
+            type as float32. Default is 0.9.
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: Add a example to show how to do decay for momentum-var

Comment (heavengate) [reply python/paddle/fluid/layers/nn.py]: Done, add example below

File: paddle/fluid/operators/batch_norm_op.cc (hunk 1 above)
Comment (heavengate) [reply paddle/fluid/operators/batch_norm_op.cc]: Done, Thanks!

File: paddle/fluid/operators/batch_norm_op.cu (hunk 2 above)
Comment (heavengate) [reply paddle/fluid/operators/batch_norm_op.cu]: Done, Thanks!

File: python/paddle/fluid/layers/nn.py (hunk 3 above)
Comment (heavengate) [reply python/paddle/fluid/layers/nn.py]: Done, Thanks!

File: python/paddle/fluid/layers/nn.py [hunk 5]
```diff
@@ -2487,6 +2488,31 @@ def batch_norm(input,
             x = fluid.data(name='x', shape=[3, 7, 3, 7], dtype='float32')
             hidden1 = fluid.layers.fc(input=x, size=200, param_attr='fc1.w')
             hidden2 = fluid.layers.batch_norm(input=hidden1)
+
+            # batch_norm with momentum as Variable
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py:2494]: split into two examples.

Comment (heavengate) [reply python/paddle/fluid/layers/nn.py:2494]: Done, Thanks!

**Review by qingqing01 — DISMISSED** [review]

**Review by wanghaoshuang — DISMISSED** [review]

**Review by guoshengCS — DISMISSED** [review]

**Review by wanghaoshuang — APPROVED** [review]

**Review by saxon-zh — APPROVED** [review]
LGTM

**Pull request merged**